ingestion-beam/src/main/java/com/mozilla/telemetry/options/OutputType.java

```diff
@@ -164,56 +166,64 @@
-          .apply(LimitPayloadSize.toMB(1));
-
-      ResultWithErrors<PCollection<KV<TableDestination, TableRow>>> tableRows = sizeLimited.output()
-          .apply(PubsubMessageToTableRow.of(tableSpecTemplate));
-
-      WriteResult writeResult = tableRows.output().apply(BigQueryIO //
+      BigQueryIO.Write<KV<TableDestination, TableRow>> writeTransform = BigQueryIO //
           .<KV<TableDestination, TableRow>>write() //
           .withFormatFunction(KV::getValue) //
           .to((ValueInSingleWindow<KV<TableDestination, TableRow>> vsw) -> vsw.getValue().getKey())
-          .withMethod(BigQueryIO.Write.Method.STREAMING_INSERTS) //
+          .withMethod(writeMethod.method)
           .withCreateDisposition(BigQueryIO.Write.CreateDisposition.CREATE_NEVER) //
           .withWriteDisposition(BigQueryIO.Write.WriteDisposition.WRITE_APPEND) //
-          .skipInvalidRows() //
-          .ignoreUnknownValues() //
-          .withExtendedErrorInfo()
-          .withFailedInsertRetryPolicy(InsertRetryPolicy.retryTransientErrors()));
-
-      PCollection<PubsubMessage> failedInserts = writeResult.getFailedInsertsWithErr()
-          .apply("Process failed inserts", MapElements.into(TypeDescriptor.of(PubsubMessage.class))
-              .via((BigQueryInsertError bqie) -> {
-                Map<String, String> attributes = new HashMap<>();
-                attributes.put("error_type", "failed_insert");
-                attributes.put("error_table",
-                    String.format("%s:%s.%s", bqie.getTable().getProjectId(),
-                        bqie.getTable().getDatasetId(), bqie.getTable().getTableId()));
-                if (!bqie.getError().getErrors().isEmpty()) {
-                  // We pull out the first error to top-level attributes.
-                  ErrorProto errorProto = bqie.getError().getErrors().get(0);
-                  attributes.put("error_message", errorProto.getMessage());
-                  attributes.put("error_location", errorProto.getLocation());
-                  attributes.put("error_reason", errorProto.getReason());
-                }
-                if (bqie.getError().getErrors().size() > 1) {
-                  // If there are additional errors, we include the entire JSON response.
-                  attributes.put("insert_errors", bqie.getError().toString());
-                }
-                byte[] payload = bqie.getRow().toString().getBytes();
-                return new PubsubMessage(payload, attributes);
-              }));
+          .ignoreUnknownValues();
+
+      if (writeMethod == BigQueryWriteMethod.streaming) {
+        writeTransform = writeTransform
+            .withFailedInsertRetryPolicy(InsertRetryPolicy.retryTransientErrors()).skipInvalidRows() //
+            .withExtendedErrorInfo();
+      } else if (writeMethod == BigQueryWriteMethod.file_loads) {
+        if (inputType == InputType.pubsub) {
+          // Trigger frequency must be set iff we're in streaming mode.
```

Comment:
```suggestion
          // Trigger frequency must be set iff we're in batch mode.
```

Reply:
Nomenclature here gets iffy here, so I'll clarify the comment. Trigger frequency must be set if and only if the overall pipeline is in streaming mode (due to pubsub input) and the bigquery output is in file_loads mode.

Reply:
oooh, okay, i see now